### DISCUSSION TITLE
tutorials/physics/using_character_body_2d

### DISCUSSION THREAD

**Discussion by giscus** [godotengine/godot-docs-user-notes#44] [User-contributed notes]
# tutorials/physics/using_character_body_2d

Introduction: Godot offers several collision objects to provide both collision detection and response. Trying to decide which one to use for your project can be confusing. You can avoid problems an...

https://docs.godotengine.org/en/stable/tutorials/physics/using_character_body_2d.html

**Comment by bazylevnik0** [comment]
Which way is the best to implement "sync to physics" in CharacterBody3D? When needed "to attach" player to moving platform

**Reply by ivandardi** [reply]
I'd try using RemoteTransform2D. Anytime you detect that a player is on a platform, set the remote transform to the platform. It should make the character stay on the platform but also be able to move while on it.

**Reply by bazylevnik0** [reply]
Ye good method thx. But when complex transforms: for example player has own rotation and platform has own rotation, it is hard to synchronize them. But thanks!

**Reply by bazylevnik0** [reply]
Found one more solution: use for platform AnimatableBody3D and when move platform by animation possible to switch sync_on_physics

**Comment by LordMcMutton** [comment]
Am I missing where the documentation details how CharacterBody decides which CollisionShape to use? 

Does it use every single one? Does it use only one? If the latter, which one does it use? Does it only use direct children? Or does it use subchildren, too?

**Reply by Calinou** [reply]
It uses all *direct* collision shape children, like other PhysicsBody type. Indirect children are currently ignored (see https://github.com/godotengine/godot-proposals/issues/535).

**Reply by LordMcMutton** [reply]
Thanks for responding!

Is this mentioned anywhere in the documentation? I can make a proposal to add it in if not, though perhaps your reply suffices?

**Reply by Calinou** [reply]
Pull request opened:

- https://github.com/godotengine/godot-docs/pull/10486

**Comment by ericsartor** [comment]
In the last example, gravity is applied using the physics delta, but the actual movement is applied without the physics delta, only the speed multiplier.

Is this because internally, move_and_slide and move_and_collide use the physics delta on the velocity? I'm unclear on how delta plays into the move functions specifically. If I was doing position based movement (like in _update), I would have to use the frame delta to ensure I move the right amount, and I think I'm having a hard time understanding why I don't have to do the same thing to the velocity vector for movement in _physics_update.

**Reply by LordMcMutton** [reply]
I believe that Velocity computation already involves the time component, so delta isn't necessary

CharacterBody's Velocity says:
"Current velocity vector (typically meters per second)"

**Reply by HVallad** [reply]
I dont know if this was figured out off this chain, but its moreso due to how the value is being set per frame. 

For velocity.x = direction * speed, its the same value, so no matter how many times the method is called per second, its direction * speed.
For velocity.y += gravity * delta, the change is additive, so the frame delta actually needs to play a role because if the computer is faster than the rest, itll be called more and therefore accelerate faster. the delta breaks the gravity acceleration up and regulates it over the frames.

Hope that helps clear up any confusion!

EDIT: To put it another way, for x you were hard setting the velocity, but for y you were adding in the acceleration to what ever velocity existed.

**Comment by RayKoopa** [comment]
Bullet collisions in the "Bouncing/reflecting" paragraph are not "resolved" entirely - only the new velocity is calculated to make them move into the new direction in the next physics integration, glueing the bullet to the wall / collision point for one integration step. Any travel remaining after the bullet hit the wall is not applied. Maybe this should be noted?

Code example to add the remaining travel (not optimized, just to point out the issue):
```gdscript

func _physics_process(delta: float) -> void:
	var remain := velocity * delta
	while not remain.is_zero_approx(): # likely want to allow larger values
		var collision := move_and_collide(remain)
		if not collision:
			break

		remain -= collision.get_travel()
		remain = remain.bounce(collision.get_normal())
		velocity = velocity.bounce(collision.get_normal()) # likely want to move this out of the loop
		
		if collision.get_collider().has_method("hit"):
			collision.get_collider().hit()
```

**Comment by eobet** [comment]
From a beginner's perspective, I want to note that the .name in collision.get_collider().name does not have a popup menu highlight like pretty much everything else on this page, so in the beginning you think you are making a mistake at this point.

(I realize the technical explanation for this as it is a property, but it's still jarring for a beginner.)

**Comment by thooombs** [comment]
It would be nice to see here a scene where the player slides if he is on a cliff, like as if he is a ball.

**Reply by thooombs** [reply]
I got the desired result it by using
```gdscript
func _ready():
	floor_stop_on_slope = false
```
